cloud/src/meta-service/meta_service_txn.cpp

```diff
@@ -913,320 +1458,420 @@ void commit_txn_immediately(
+            int64_t table_id = tablet_ids[tablet_id].table_id();
+            int64_t partition_id = i.partition_id();
+            std::string ver_key =
+                    partition_version_key({instance_id, db_id, table_id, partition_id});
+            if (new_versions.count(ver_key) == 0) {
+                new_versions.insert({ver_key, 0});
+                version_keys.push_back(std::move(ver_key));
+            }
+        }
+
+        std::vector<std::optional<std::string>> version_values;
+        err = txn->batch_get(&version_values, version_keys);
+        if (err != TxnErrorCode::TXN_OK) {
+            code = cast_as<ErrCategory::READ>(err);
+            ss << "failed to get partition versions, err=" << err;
             msg = ss.str();
+            LOG(WARNING) << msg << " txn_id=" << txn_id;
             return;
         }
-        rowsets.emplace_back(std::move(key), std::move(val));
 
-        // Accumulate affected rows
-        auto& stats = tablet_stats[tablet_id];
-        stats.data_size += i.data_disk_size();
-        stats.num_rows += i.num_rows();
-        ++stats.num_rowsets;
-        stats.num_segs += i.num_segments();
-    } // for tmp_rowsets_meta
+        for (size_t i = 0; i < version_keys.size(); i++) {
+            int64_t version;
+            if (version_values[i].has_value()) {
+                VersionPB version_pb;
+                if (!version_pb.ParseFromString(version_values[i].value())) {
+                    code = MetaServiceCode::PROTOBUF_PARSE_ERR;
+                    ss << "failed to parse version pb txn_id=" << txn_id
+                       << " key=" << hex(version_keys[i]);
+                    msg = ss.str();
+                    return;
+                }
+                if (version_pb.has_txn_id()) {
+                    need_advance_last_txn = true;
+                    last_pending_txn_id = version_pb.txn_id();
+                    break;
+                }
+                version = version_pb.version();
+            } else {
+                version = 1;
+            }
+            new_versions[version_keys[i]] = version + 1;
+            need_advance_last_txn = false;
+            last_pending_txn_id = 0;
+        }
+
+        if (need_advance_last_txn) {
+            txn.reset();
+            DCHECK(last_pending_txn_id > 0);
+            std::shared_ptr<TxnLazyCommitTask> task =
+                    txn_lazy_committer->submit(instance_id, last_pending_txn_id);
+
+            std::pair<MetaServiceCode, std::string> ret = task->wait();
+            need_advance_last_txn = false;
+            last_pending_txn_id = 0;
+            code = ret.first;
+            msg = ret.second;
+            if (code != MetaServiceCode::OK) {
+                LOG(WARNING) << "advance_last_txn failed last_txn=" << last_pending_txn_id
+                             << " code=" << code << "msg=" << msg;
+                return;
+            }
+            continue;
```

Comment:
Add some comments to explain why we need to retry.

Reply:
> Add some comments to explain why we need to retry.

done, i have added a comment